core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -19,9 +24,51 @@
 import io.reactivex.rxjava3.core.Flowable;
 
 /**
+ * The contract for defining a way to interface with external source of data, such as a database, filesystem etc. As
+ * the name implies, all of the methods in this class must <b>never</b> block the invoking thread.
  * <p>
- * Implementations of this store must be thread safe if concurrent operations are performed on it. This should include
- * possibly invoking start or stop multiple times
+ * The first method that will be invoked on this store will be the {@link #start(InitializationContext)} to allow
+ * it to initialize and startup. Once the returned stage has completed the store is assumed to be in working condition
+ * for other operations to be performed upon it. Infinispan will guarantee visibility of variables written during
+ * the start method, so there is no need to synchronize these manually, unless they are mutated in the normal operations
+ * of the store itself.
+ * <p>
+ * Infinispan may at any point after this store has started query what characteristics this store has via
```

Comment:
"After the store has started, Infinispan will utilise the {@link #characteristics()} method to query the store's characteristics."